### PR TITLE
tpm2_create: add TPM CreateLoaded support

### PR DESCRIPTION
**Pull request by williamcroberts**
Allow tpm2_create to take the -c/--context output option to
produce a context file similar to tpm2_load()

Internally, this will yeild a call to ESYS_CreateLoaded over
ESYS_Create thus producing the loaded ctx object.

Fixes: #832

Signed-off-by: William Roberts <william.c.roberts@intel.com>

### CODE REVIEW

File: tools/tpm2_create.c [hunk 1]
```diff
@@ -119,39 +119,95 @@ static bool create(ESYS_CONTEXT *ectx) {
...
+            ESYS_TR parentHandle,
+            ESYS_TR shandle1,
+            ESYS_TR shandle2,
+            ESYS_TR shandle3,
+            const TPM2B_SENSITIVE_CREATE *inSensitive,
+            const TPM2B_TEMPLATE *inPublic,
+            ESYS_TR *objectHandle,
+            TPM2B_PRIVATE **outPrivate,
+            TPM2B_PUBLIC **outPublic);
+         */
```
Comment (williamcroberts) [inline tools/tpm2_create.c]: whoops, delete me

File: tools/tpm2_create.c [hunk 2]
```diff
@@ -220,9 +266,10 @@ bool tpm2_tool_onstart(tpm2_options **opts) {
       { "pubfile",              required_argument, NULL, 'u' },
       { "privfile",             required_argument, NULL, 'r' },
       { "context-parent",       required_argument, NULL, 'C' },
+      { "context",              required_argument, NULL, 'c' },
```
Comment (joshuagl) [inline tools/tpm2_create.c]: `-c`/`--context` is by convention a child context for a command, would this be better as `-o`/`--out-context` like `tpm2_load` ?

File: tools/tpm2_create.c [hunk 3]
```diff
@@ -119,39 +119,82 @@ static bool create(ESYS_CONTEXT *ectx) {
...
+        if(rval != TPM2_RC_SUCCESS) {
+            LOG_PERR(Esys_Create, rval);
+            return false;
+        }
+        free(creationData);
+        free(creationHash);
+        free(creationTicket);
     }
 
+
```
Comment (joshuagl) [inline tools/tpm2_create.c]: Spurious whitespace?

File: tools/tpm2_create.c [hunk 4]
```diff
@@ -119,39 +119,82 @@ static bool create(ESYS_CONTEXT *ectx) {
...
-    rval = Esys_Create(ectx, ctx.context_object.tr_handle,
-            shandle1, ESYS_TR_NONE, ESYS_TR_NONE,
-            &ctx.in_sensitive, &ctx.in_public, &outsideInfo, &creationPCR,
-            &outPrivate, &outPublic, &creationData, &creationHash,
-            &creationTicket);
-    if(rval != TPM2_RC_SUCCESS) {
-        LOG_PERR(Esys_Create, rval);
-        ret = false;
-        goto out;
+    ESYS_TR object_handle;
```
Comment (joshuagl) [inline tools/tpm2_create.c]: Initialise this to ESYS_TR_NONE to placate scan-build?

File: tools/tpm2_create.c (hunk 2 above)
Comment (williamcroberts) [reply tools/tpm2_create.c]: Yeah I thought about this as well, initially  I wasn't happy with -o option on tpm2_load but considering you can load a child or parent, it makes more sense.

Comment (williamcroberts) [reply tools/tpm2_create.c]: Actually, -c is really "this object context". For instance tpm2_readpublic, where as -C is this objects parent.

Comment (williamcroberts) [reply tools/tpm2_create.c]: Ill keep it same as tpm2_load and we can figure out the proper ones and be consistent when we go through all the options for release.

File: man/tpm2_create.1.md [hunk 5]
```diff
@@ -90,10 +97,19 @@ These options for creating the tpm entity:
...
+```
+tpm2_create -C parent.ctx  -K def456 -G keyedhash -I seal.dat -u obj.pub -r obj.priv
+```
+
+Create an rsa2048 object and load it into the TPM:
 ```
-tpm2_create -C 0x81010001 -P abc123 -K def456 -g sha256 -G keyedhash-I data.File
-tpm2_create -C file:parent.context -P abc123 -p def456 -g sha256 -G keyedhash -I data.File
-tpm2_create -C 0x81010001 -P 123abc -K 456def -X -g sha256 -G keyedhash -I data.File
+tpm2_create -C primary.ctx -G rsa2048 -u obj.pub -r obj.priv -c obj.ctx
```
Comment (joshuagl) [inline man/tpm2_create.1.md]: Should be -o obj.ctx now

Comment (williamcroberts) [reply man/tpm2_create.1.md]: Good catch... these always seem to miss being updated

**Comment by williamcroberts** [comment]
> Hmm, tpm2_readpublic call failed in create.sh integration test.

Helps if you change the case statement to 'o' from 'c' as well.

**Pull request merged**